### DISCUSSION TITLE
Pyinstaller problem with a module that uses torch and pytorch_lightning.

### DISCUSSION THREAD

**Discussion by Caio-Giulio-Cesare** [pyinstaller/pyinstaller#7944] [Help]
Hi everyone, I can't create an executable with pyinstaller when I use a module that itself uses torch and pytorch_lightning.

The script ( main.py ) basically does nothing except import that module.

The first problem occurs during the creation phase, but this problem, as suggested by the creation phase, is resolved by adding the following line to the main.spec file

_import sys ; sys.setrecursionlimit(sys.getrecursionlimit() * 5)_

and launching the command

_pyinstaller main.spec_

But even in this case, although the executable is created, it does not work.

I attach the exe_error_.jpg file with the error details.

Even trying to create the executable not in "onefile mode" the resulting executable does not work.

I tried to apply the indications found in previous discussions but I didn't solve it.

Has this problem happened to anyone?
Is there a way to fix and create the executable?

Thanks to anyone who wants to help me.

![exe_error_](https://github.com/pyinstaller/pyinstaller/assets/84173400/43eaa081-4aea-46ac-9595-0ad5fac68634)
![script_main_py](https://github.com/pyinstaller/pyinstaller/assets/84173400/a6e48f2c-da45-47f4-ac3b-b089adcd1f4b)

**Comment by rokm** [comment]
What version of PyInstaller and pyinstaller-hooks-contrib are you using?

**Reply by Caio-Giulio-Cesare** [reply]
Hi rokm, thanks for the reply.

I use these modules

pyinstaller 5.13.2

pyinstaller-hooks-contrib 2022.5

pyinstaller-versionfile 2.1.1

Thanks again

**Reply by rokm** [reply]
Try updating `pyinstaller-hooks-contrib` to the latest version (`2023.8`) - you would need at least `2022.9` for `cv2` to work properly (your current issue is not `torch`/`pytorch_lightning`, but `cv2`).

**Reply by Caio-Giulio-Cesare** [reply]
Hi rokm, thanks again for your kindness.

I have updated all the modules with the latest versions shown below.
_pyinstaller 5.13.2
pyinstaller-hooks-contrib 2023.8
pyinstaller-versionfile 2.1.1_

The situation has improved but not yet completely resolved.

When I launch the executable it tells me that it cannot find the _rasterio.sample module_ (see image of the exe_error_2.jpg log)

Yet I installed this module (see image detail_rasterio_sample.png).

Can you think of how I can solve this?

Thank you very much again, and I hope to be able to return the favor one day.

![exe_error_2](https://github.com/pyinstaller/pyinstaller/assets/84173400/2cf07a8e-da6c-4755-a3e7-d90197dab947)
![detail_rasterio_sample](https://github.com/pyinstaller/pyinstaller/assets/84173400/87b3bf54-2830-45ce-8584-0792368b0eb8)

**Reply by rokm** [reply]
Add the missing module name (`rasterio.sample`) to `hiddenimports` list in the spec file. (It is not automatically picked up because it is imported from `rasterio._io`, which is a cythonized python extension, and PyInstaller cannot analyze those for imports).

If you get more missing modules from `rasterio`, you can add `from PyInstaller.utils.hooks import collect_submodules` at the top of the spec, and then do `hiddenimports = collect_submodules('rasterio')` to force collection of all submodules from that package.

**Reply by Caio-Giulio-Cesare** [reply]
Works! Thank you so much and I have no words to thank you!

Now the executable also works, I only get some warnings but I hope they aren't a blocking problem.

Thank you very much again, tell me how I can reciprocate your kindness.

![only_warnings](https://github.com/pyinstaller/pyinstaller/assets/84173400/bc5780fb-2927-4453-bcfe-6ea334996858)

**Reply by rokm** [reply]
Those warnings are coming from `torchvision`. The first one is because an extension is not collected - this can be remedied by doing a `collect_submodules('torchvision')`. The warnings about JIT and `_DenseLayer.forward` are caused by lack of source .py files for JIT (as PyInstaller does not collect source .py files by default).

Modifying your spec as follows should take care of those warnings:

```python
from PyInstaller.utils.hooks import collect_submodules

hiddenimports = []
hiddenimports += collect_submodules('rasterio')
hiddenimports += collect_submodules('torchvision')

a = Analysis(
    ...
    hiddenimports=hiddenimports,
    ...
    module_collection_mode = {
        'torchvision': 'pyz+py',  # Ensure that source .py files are collected (in addition to .pyc modules in PYZ archive).
    },
)

...
```

**Reply by Caio-Giulio-Cesare** [reply]
It works perfectly! With this latest change the warnings have also disappeared.

Thank you very much again, you were very kind.
